### DISCUSSION TITLE
Better syntactic contexts

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#8534] [Ideas]
A syntactic context is almost just a call stack of the parser, so why not use that?

At the start of parseEffect, have `implicit context = Context.Decl.Effect`. Then if the code goes to `parseExp` then it'll override `implicit context = Context.Exp.OtherExp` or whatever.

**Comment by magnus-madsen** [comment]
I think that is more or less what we are doing? What are the concrete things you want to be different?

**Reply by JonathanStarup** [reply]
This snippet fx keeps reiterating that it is `SyntacticContext.Expr.OtherExpr`
```scala
    private def catchRule()(implicit s: State): Mark.Closed = {
      val mark = open()
      expect(TokenKind.KeywordCase, SyntacticContext.Expr.OtherExpr)
      name(NAME_VARIABLE, context = SyntacticContext.Expr.OtherExpr)
      expect(TokenKind.Colon, SyntacticContext.Expr.OtherExpr)
      name(NAME_JAVA, tail = Set(), allowQualified = true, context = SyntacticContext.Expr.OtherExpr)
      expect(TokenKind.ArrowThickR, SyntacticContext.Expr.OtherExpr)
      statement()
      close(mark, TreeKind.Expr.TryCatchRuleFragment)
    }
```
In the implicit world it would be
```scala
    private def catchRule()(implicit s: State): Mark.Closed = {
     implicit val context: SyntacticContext = SyntacticContext.Expr.OtherExpr
      val mark = open()
      expect(TokenKind.KeywordCase)
      name(NAME_VARIABLE)
      expect(TokenKind.Colon)
      name(NAME_JAVA, tail = Set(), allowQualified = true)
      expect(TokenKind.ArrowThickR)
      statement()
      close(mark, TreeKind.Expr.TryCatchRuleFragment)
    }
```
Reading is easier because distinct constructs have a def which initially says "I am XYZ context". An upside is that is also becomes clear which functions inherit the context and which functions redefine their own context.

**Reply by magnus-madsen** [reply]
I am not opposed to that...

Unless, do you have examples where there are many different contexts in play?

**Reply by JonathanStarup** [reply]
There is 328 contexts but skimming through I don't think so. Lots of the code is like this, like def instance just saying that it is an instance
```scala
    private def instanceDecl(mark: Mark.Opened)(implicit s: State): Mark.Closed = {
      assert(at(TokenKind.KeywordInstance))
      expect(TokenKind.KeywordInstance, SyntacticContext.Decl.Instance)
      name(NAME_DEFINITION, allowQualified = true, context = SyntacticContext.Decl.Instance)
      if (!eat(TokenKind.BracketL)) {
        // Produce an error for missing type parameter.
        expect(TokenKind.BracketL, SyntacticContext.Decl.Instance, hint = Some("Instances must have a type parameter."))
      } else {
        Type.ttype()
        expect(TokenKind.BracketR, SyntacticContext.Decl.Instance)
      }
      ...
```

**Reply by magnus-madsen** [reply]
So the concrete proposal is: 1. make argument implicit 2. define (or redefine) in every rule?

**Reply by JonathanStarup** [reply]
yes exactly

**Reply by magnus-madsen** [reply]
Lets discuss on monday.
